rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,276 @@
+---
+authors: Alexander Klizhentas (sasha@goteleport.com), Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 59 - Search Based Access Requests
+
+## What
+
+This RFD proposes just-in-time search-based access requests which will enable
+users to request access to resources based on a search and/or selection of
+individual resources, rather than having to request one or more roles.
+
+## Why
+
+Teleport users may not know in advance which roles they need or who should grant
+access, just which resource(s) they need access to.
+
+Users would like to avoid having many different custom roles for different
+levels of access.
+
+Requesting access to a "blanket" role violates the principle of least priviledge.
+Most of the time incident responders only need access to 1 or 2 nodes,
+search-based access requests will allow them to request access to only the nodes
+they need.
+
+
+## Details
+
+Users will be able to request access by using search, and roles will be used
+behind the scenes to evaluate who will be responsible for reviewing and granting
+access requests.
+
+Users won’t be asked to think about roles in advance, but will be able to focus
+on finding and requesting access to individual resources.
+
+The existing “role-based” access requests and the new “search-based” access
+requests proposed here are useful for two different scenarios:
+
+- Sometimes it is helpful to request elevated roles in the system, like
+  `db-admin`, to perform some system upgrades, or system wide-troubleshooting.
+  In this case, role-based access requests work great.
+
+- In other cases, it’s helpful to request access to a subset of resources. In
+  this case, search.
+
+### Example Flow
+
+Alice is a member of the group “splunk”.
+
+She gets an alert that “db-1” is malfunctioning.
+
+Alice needs to locate the "db-1" resource and request access to it.
+
+Alice goes to the “Request access” screen and selects: “select access to resources”
```

Comment:
We would really like that the API is simply that Alice does 'tsh ssh db-1' and is told that she needs to give a reason and wait for permission from Slack group X. The vast majority of our users do not use the Teleport web UI (except as a clickable bounce to MFA auth) - they use their terminal. 

tsh can of course do this lookup via API, e.g. figure out that Alice is not a member of a role that has normal access to db-1, but she is in this group that can access it via permission. But from the users POV its all the same - some hosts they can get into, some hosts they cant, and there is this new middle ground. 

To be clear I think its fine for this to only work for 'tsh ssh' (rather than openssh), certainly initially.  If it can work in the web UI as well, thats great, but thats less important for us than the tsh command.

Reply:
We will make this work with `tsh ssh` as mentioned in other comments and in the doc now.